### ISSUE TITLE
to_word()使训练变得无意义

### ISSUE THREAD

**Issue by vpegasus** [jinfagang/tensorflow_poems#19]
你好,谢谢你提供的demo, 不知你是否就是'斗大的熊猫', 反正是真的谢谢了.

我认真的读了原码,发现在生成古诗的函数中,有一个子函数: to_word(). 这是一个随机函数啊, 正常应该是取argmax()啊, 但我修改后,发现进入死循环.
应用to_word()确实可以避开死循环,不过因为是随机,也使得训练变得没有意义了~

**Comment by gaussic** [comment]
个人同样认为并没有太大的意义。不过可能是因为训练的数据集对于RNN来说还是比较小，因此在生成时经常出现",,,,,,,......."这样的情况。

类似的模型，我用三国演义的60万字语料库，基于PYTORCH做的版本，感觉效果还值得一看：

> 王中、潘充士列曰：“母蒋忠来行，何在取此也！”不想孔融保程泰于马前蛮仁谏。邓虔大骂，“云：“誓来二年，军马不意，如何言我耶？”近都立下乘车上看见曹操，乃谓首县一枪铜。

> 彭丹伍，不开出中锋，乃有不归，还治武安人，报是丞相诣寰之，司马昭入帐痛贼，留刘皇外引而退。正是夜中张郃乱起，袍酗金白俟巾；急于一千军箭甲火声，刀辆昧掩，伤岸插无二野之战。孔明答于合下，率兵出大路。

> 随诏大王，观荆州呆而下，虔如吾获，皆准君。当日作闻明，乃与大军、张辽复至河寨！看知全命，却说关长玲营来来，却彼领兵归后大散。正约魏延正往荆襄而江；幡内八半，尽跃可进，内地相赴。云长领军奔取秦策。

> 于望为弃之，许弟会处擂身不勇，数日山劳。云长就欲先踏也。须观，再不过；云长进在在宗庶，随此不能辞读。楙拜作内相侍，手措苦为，瑶剑闭乱，皆回女太守使王，操与刘达宁事、高常救慰。

> 令将军分住；汝必来断，彼闻必归军士璆深发。就见鲁权窗锐羌洞者城去：运鹤远处。儒与御内俱下一军上环令而从。又将船去投荆州与都引兵到，凌德羌主拆于荆州，骁半抬手弓持一齐，不敢杀疑。那部抄埋起起，急飞乘枪

详见[Language Model Zh](https://github.com/gaussic/language_model_zh/blob/master/lm_chinese.ipynb)

**Comment by jinfagang** [comment]
@vpegasus I am not him, but I borrowed some code from him, the implementation still have some variance. @gaussic  This is also a good job. If you using PyTorch RNN tutorial model to generate sentences the model should be not complicated enough. A father more experiment on GAN models to generate article still under researching due to my upload on finding a job, I will open source it as soon as get some break through.

**Comment by chinaos** [comment]
解释了to_word()这个函数的作用了吗？

@vpegasus  @gaussic

**Comment by zl-nemo** [comment]
个人觉得to_word()是有意义的，sample = int(np.searchsorted(t, np.random.rand(1) * s))虽然带有一定的随机性，但predict经模型预测出，并非随机变量，所以t = np.cumsum(predict)也服从某种分布，但不会是均匀分布的累计分布，因此sample也是大概率地落入到argmax(predict)对应的位置。作者其实用了一种巧妙的方式同时规避了死循环和保留predict意义的方式。 @vpegasus @gaussic @chinaos

**Comment by gaussic** [comment]
以下面这段代码为例：

```python
predict = np.array([0.1, 0.5, 0.1, 0.1, 0.2])   # 满足softmax
t = np.cumsum(predict)
print('t:', t)
```

```
t: [ 0.1  0.6  0.7  0.8  1. ]
```

可以看到，`cumsum()`其实就是对输出做累积。

```python
s = np.sum(predict)
print('s:', s)
```

```
s: 1.0
```

在模型中就已经取了`tf.nn.softmax()`了，所以我觉得这句是没有意义的。

```python
sample = np.searchsorted(t, np.random.rand(1) * s)
print('sample:', sample)
```

```
sample: [3]
```

再看，`searchsorted()`的含义，是找出后面生成的那个随机数，应该插入到t中的哪个位置。

接下来我们来测试抽样100000次：

```python
cnt = np.zeros(len(predict))
for _ in range(100000):
    sample = np.searchsorted(t, np.random.rand(1) * s)[0]
    cnt[sample] += 1
print(cnt)
```

```
[  9913.  50112.  10034.   9874.  20067.]
```

100000次抽样，其在各个词上的概率与predict接近。也就是说，这段代码应该是服从多项式分布的。所以这个`to_word()`的目的是使生成的结果并不局限于生成概率最高的那个词，而是在topK之间波动。

不过，我才po主本身也没有太过理解这一段代码。



因为，大部分的code是从[TensorFlow练习7: 基于RNN生成古诗词](http://blog.topspeedsnail.com/archives/10542)

**借**

**来**

**的**

对原文没有做任何的引用，并加上自己的Lincence，私以为是**不太道德**的。

@vpegasus @chinaos @wossoneri @zl-nemo
